README.md
interSubs
=========

Interactive subtitles for [mpv](https://github.com/mpv-player/mpv), that was made to help study languages.  
Easily tweaked and customizable.

![z 00_00_5 75-00_00_19 96](https://user-images.githubusercontent.com/10230453/38359595-7f56acc0-38d1-11e8-9a65-257466a44e08.gif)

* Supported dictionaries for words:
	* https://dict.cc/
	* https://pons.com/
	* http://reverso.net/
	* https://dict.leo.org/
	* https://translate.google.com/
	* http://morfix.co.il/
	* https://redensarten-index.de/
* Supported dictionaries for sentences:
	* https://www.deepl.com/translator
	* https://translate.google.com/
* Supported dictionaries for word pronunciation:
	* https://forvo.com/
	* https://pons.com/
	* https://translate.google.com/
* Offline \t separated dictionary. [details](https://github.com/oltodosel/interSubs/issues/36#issuecomment-803541985) | [pyglossary](https://github.com/ilius/pyglossary)
* http://linguee.com/ redirecting to browser by click.
* Can use multiple dictionaries simultaneously.
* Reassigning mouse buttons functions in config.
* Doesn't work with DVD (picture based) subtitles, only the text-based ones.
	* [Script](https://github.com/oltodosel/extract_n_convert_dvd_bd_subtitles) to convert picture based subtitles into *.srt; also extracts them from *.mkv 
* Can extend time of subs showing; for slow readers
```
    00:02:23,046 --> 00:02:25,990
    bla bla
    00:02:28,020 --> 00:02:33,084
    waka waka
    
    00:02:23,046 --> 00:02:28,020
    bla bla
    00:02:28,020 --> 00:02:33,084
    waka waka
```

Requirements
------------
   * mpv 0.27 (I don't know if it will work with mpv front-ends.)
   * Xorg (ignore for Mac users)
   * composite manager; `xcompmgr` or sth. (ignore for Mac users)
   * python => 3.6
   * python-pyqt5
   * python-numpy
   * python-beautifulsoup4
   * python-requests
   * python-lxml
   * python-httpx
   * lua
   * socat
   * pkill
   * xdotool (for hiding subtitles when minimizing mpv or switching window) 
   * optional: chromium (for external translation, other browser can be specified)
   * optional: wget (for listening to pronunciation)

Installation
------------
* `mv interSubs.py interSubs_config.py ~/.config/mpv/scripts/; mv main.lua ~/.config/mpv/scripts/interSubs.lua`
  * or `cd ~/.config/mpv/scripts/; git clone "this repository link"`
* Edit configuration file `interSubs_config.py`
* Edit the lua file if you want change keybindings and should the plugin auto-start or not. 
* For Mac also edit configuration at `interSubs.lua`
* On KDE(kwin) go to Compositor and uncheck "Allow applications to block compositing". [Screenshot](https://iwf1.com/wordpress/wp-content/uploads/2017/09/Disable-applications-override-compositor-KDE.jpg).
* For Windows - port it yourself. (It does NOT work on Windows. [reason](https://github.com/mpv-player/mpv/blob/master/DOCS/man/ipc.rst#command-prompt-example)) and [a possible way to do it](https://github.com/oltodosel/interSubs/issues/40#issuecomment-835767040)). Consider installing VirtualBox and Ubuntu or some other user-friendly distro.

Usage
-----
* Start video with mpv & select subtitles.
* F5 to start/stop interSubs.
* Point cursor over the word to get popup with translation.
* F6 to hide/show without exiting.

Buttons bellow may be reassigned in config
-----
* Left-click - show translation in your browser.
* Right-click - listen to pronunciation.
* Wheel - scroll through transitions in popup.
* Wheel+Ctrl - resize subtitles.
* Wheel+Shift - change subtitles' vertical position.
* Wheel-click - cycle through auto_pause options.
* Wheel-click-left/right - +/- auto_pause_min_words. (fancy mouses)
* Back-click - translate whole sentence. (fancy mouses)

Troubleshooting
-----
* Does not work at all - launch `mpv` in a console, then start `interSubs` and see if there are any errors that might indicate on missing packages and install them.
* Background and subtitles are shown as solid black bars - install/launch a compositor.
* May have issues working in a multi-monitor system.  See [the solution](https://github.com/oltodosel/interSubs/issues/26).
* On KDE subtitles might sometimes be invisible. See [the solution](https://github.com/oltodosel/interSubs/issues/12#issuecomment-433960146).
* On KDE subtitles fade instead of appearing immediately. See [the solution](https://github.com/oltodosel/interSubs/issues/39#issuecomment-810483614)
* Instead of changing system settings you may change [--x11-bypass-compositor](https://mpv.io/manual/stable/#options-x11-bypass-compositor)
* Stuttering video during subtitles change might be solved by changing mpv's video output `mpv --vo gpu`.
* On MacOS subtitles are not rendered on top of `mpv`. See [a possible solution](https://github.com/oltodosel/interSubs/issues/42#issuecomment-841725012)
* On MacOS subtitles are not rendered on top of `mpv` only in fullscreen mode. See [the solution](https://github.com/oltodosel/interSubs/issues/42#issuecomment-841746404)

[Changelog](CHANGELOG.md)
-----

CHANGELOG.md
Changelog
-----
* 2.0a
	* Configs are incompatible with previous version.
	* Tk is abandoned in favor of Qt.
		* Background can be fully transparent or semi/fully opaque.
		* Rendering is faster than with Tk.
		* requires pyqt5 for python 3
			* `pip install pyqt5` / `pacman -S python-pyqt5`
		* requires composite manager for not solid bg; `xcompmgr` or sth.
		* tested on Openbox, i3, KDE(kwin).
		* On KDE(kwin) go to composite and uncheck "Allow applications to block compositing". [screenshot](https://iwf1.com/wordpress/wp-content/uploads/2017/09/Disable-applications-override-compositor-KDE.jpg)
	* No more stalling when pointing on a wrong word; those words will be translated and saved in background.
	* R2L isn't ready yet.
	* Option to not save translations on the disk was removed.
	* Noun colorization was removed.
	* Randomization of translations was removed.
	* Option to show N of previous subtitles is suspended for now, I might add it in the future.
	* Tk version won't be updated unless something critical happens.
* 2.1
	* R2L support (checked on Hebrew; works more or less).
	* Minor corrections.
* 2.2
	* Added https://dict.leo.org/ - de<>en/es/fr/it/pl/pt/ru/zh(cn)
	* Minor corrections.

* 2.3
	* Added option to limit extension of subtitles during long scenes without talking.
		* `extend_subs_duration_limit_sec = 15`
	* Updated https://translate.google.com/
		* Now it gives complete output instead of single result.
		* `mtranslate_google` -> `google`
	* Added http://www.morfix.co.il/
		* `morfix`
	* Minor error corrections.
* 2.4
	* Fixed non-working deepl.com
	* Minor corrections.
* 2.5
	* Added option to hide/show interSubs without exiting - F6
	* Minor corrections.
* 2.6
	* Fixed inability to start after update to pyqt5 5.11.3 or sth in that area. Didn't look for version that causes it specifically.
* 2.7
	* Fixed residual flickering of previous lines during subtitles change. Began to happen at qt5.12 or so.
* 2.8
	* Fixed gtts/pons pronunciation.
	* Fixed google-translation.
* 2.9
	* Added google-translation for full sentences(default mouse Back-click):
		* `f_translation_full_sentence` to bind to mouse instead of `f_deepl_translation`
		* in config `translation_function_name_full_sentence = 'google'`
* 2.10
	* Fixed google-translation


main.lua
local PLUGIN_TOGGLE_KEYBIND = "F5"
local VISIBILITY_TOGGLE_KEYBIND = "F6"
local IS_AUTO_START_ON = false
-- for Mac change python3 to python or pythonw
local PYTHON_COMMAND = 'python3 "%s" "%s" "%s"'
local PYTHON_SCRIPT_FILENAME = "interSubs.py"

local python_script_file_path = mp.get_script_directory().."/"..PYTHON_SCRIPT_FILENAME
local random_number = math.random(11111111, 99999999)

local function put_cmd_in_bg(cmd) return cmd.." &" end

local function is_there_no_selected_subs()
	return mp.get_property("sub") == "no" or mp.get_property("sub") == "auto"
end	

local function create_mpv_socket_file()
	-- recomend to have it in tmpfs
	mpv_socket_file_path = "/tmp/mpv_socket"..'_'..random_number
	mp.set_property("input-ipc-server", mpv_socket_file_path)
end

local function destroy_mpv_socket_file()
	mp.set_property("input-ipc-server", "")
	os.execute(put_cmd_in_bg("rm "..mpv_socket_file_path))
end

local function write_sub_to_file(name, value)
	if type(value) == "string" then
		file = io.open(subs_file_path, "w")
		file:write(value)
		file:close()
	end
end

local function create_subs_file()
	-- recomend to have it in tmpfs
	subs_file_path = "/tmp/mpv_sub"..'_'..random_number
	mp.observe_property("sub-text", "string", write_sub_to_file)
end

local function destroy_subs_file()
	mp.unobserve_property(write_sub_to_file)
	os.execute(put_cmd_in_bg("rm "..subs_file_path))
end

local function save_current_subs_settings()
	sub_visibility = mp.get_property("sub-visibility")
	sub_color = mp.get_property("sub-color", "1/1/1/1")
	sub_border_color = mp.get_property("sub-border-color", "0/0/0/1")
	sub_shadow_color = mp.get_property("sub-shadow-color", "0/0/0/1")
end

local function restore_subs_settings()
	mp.set_property("sub-visibility", sub_visibility)
	mp.set_property("sub-color", sub_color)
	mp.set_property("sub-border-color", sub_border_color)
	-- mp.set_property("sub-shadow-color", sub_shadow_color)
end

local function hide_native_subs()
	mp.set_property("sub-color", "0/0/0/0")
	mp.set_property("sub-border-color", "0/0/0/0")
	mp.set_property("sub-shadow-color", "0/0/0/0")
end

local function stop_intersub()
	mp.command('show-text "Quitting interSubs..."')

	os.execute(put_cmd_in_bg("pkill -f "..mpv_socket_file_path))
	destroy_mpv_socket_file()
	destroy_subs_file()
	restore_subs_settings()

	mp.unregister_event(stop_intersub)
	is_running = false
end

local function start_intersub()
	if is_there_no_selected_subs() then
		mp.command('show-text "Select subtitles before starting interSubs."')
		return 
	end
	mp.command('show-text "Starting interSubs..."')

	save_current_subs_settings()
	mp.set_property("sub-visibility", "yes")
	-- TODO: check if it needs to be like this
	mp.set_property("sub-ass-override", "force")
	hide_native_subs()

	create_subs_file()
	create_mpv_socket_file()

	os.execute(put_cmd_in_bg(PYTHON_COMMAND:format(python_script_file_path, mpv_socket_file_path, subs_file_path)))
	mp.register_event("shutdown", stop_intersub)
	is_running = true
end

local function toggle_subs_visibility()
	if not is_running then return end
	if is_visible then
		os.execute(put_cmd_in_bg('touch "'..mpv_socket_file_path..'_hide"'))
		mp.osd_message("Hiding interSubs.", .8)
		is_visible = false
	else
		os.execute(put_cmd_in_bg('rm "'..mpv_socket_file_path..'_hide"'))
		mp.osd_message("Showing interSubs.", .8)
		is_visible = true
	end
end

local function run()
	if is_running then
		stop_intersub()
	else
		start_intersub()
	end
end

local function on_file_loaded()
	if is_there_no_selected_subs() then return end
	if IS_AUTO_START_ON then
		run()
	end
end

mp.add_forced_key_binding(PLUGIN_TOGGLE_KEYBIND, "start-stop-interSubs", run)
mp.add_forced_key_binding(VISIBILITY_TOGGLE_KEYBIND, "hide-show-interSubs", toggle_subs_visibility)
mp.register_event("file-loaded", on_file_loaded)


interSubs.lua
-- v. 2.7
-- Interactive subtitles for `mpv` for language learners.
--
-- default keybinding to start/stop: F5
-- default keybinding to hide/show: F6
-- if interSubs start automatically - mpv won't show notification
--
-- dirs in which interSubs will start automatically; part of path/filename will also work; case insensitive; regexp
-- autostart_in = {'German', ' ger ', '%.ger%.', 'Deutsch', 'Hebrew'}
autostart_in = {'Hebrew'}

-- for Mac change python3 to python or pythonw
start_command = 'python3 "%s" "%s" "%s"'

-- recomend to have these in tmpfs, or at least ssd.
sub_file = '/tmp/mpv_sub'
mpv_socket = '/tmp/mpv_socket'

keybinding = 'F5'
keybinding_hide = 'F6'

pyname = '~/.config/mpv/scripts/interSubs.py'

------------------------------------------------------

debug = false
-- debug = true

if debug == true then
	start_command = ''
	start_command = 'terminator -e \'python3 "%s" "%s" "%s"; sleep 33\''
end

------------------------------------------------------

function s1()
	if running == true then
		s_rm()
		return
	end

	running = true
	mp.msg.warn('Starting interSubs ...')
	mp.register_event("end-file", s_rm)
	rnbr = math.random(11111111, 99999999)

	if debug == true then
		rnbr = ''
	end

	mpv_socket_2 = mpv_socket .. '_' .. rnbr
	sub_file_2 =  sub_file .. '_' .. rnbr
	
	-- setting up socket to control mpv
	mp.set_property("input-ipc-server", mpv_socket_2)
	
	-- without visible subs won't work
	sbv = mp.get_property("sub-visibility")
	mp.set_property("sub-visibility", "yes")
	mp.set_property("sub-ass-override", "force")
	
	sub_color1 = mp.get_property("sub-color", "1/1/1/1")
	sub_color2 = mp.get_property("sub-border-color", "0/0/0/1")
	sub_color3 = mp.get_property("sub-shadow-color", "0/0/0/1")
	mp.set_property("sub-color", "0/0/0/0")
	mp.set_property("sub-border-color", "0/0/0/0")
	mp.set_property("sub-shadow-color", "0/0/0/0")

	start_command_2 = start_command:format(pyname:gsub('~', os.getenv('HOME')), mpv_socket_2, sub_file_2)
	os.execute(start_command_2 .. ' &')

	mp.observe_property("sub-text", "string", s2)
end

function s2(name, value)
	if type(value) == "string" then
		file = io.open(sub_file_2, "w")
		file:write(value)
		file:close()
	end
end

function s_rm()
	running = false
	hidden = false
	mp.msg.warn('Quitting interSubs ...')

	mp.set_property("sub-visibility", sbv)
	mp.set_property("sub-color", sub_color1)
	mp.set_property("sub-border-color", sub_color2)
	--~ mp.set_property("sub-shadow-color", sub_color3)

	os.execute('pkill -f "' .. mpv_socket_2 .. '" &')
	os.execute('(sleep 3 && rm "' .. mpv_socket_2 .. '") &')
	os.execute('(sleep 3 && rm "' .. sub_file_2 .. '") &')
end

function started()
	if mp.get_property("sub") == 'no' then
		return true
	end
	
	hidden = false

	for kk, pp in pairs(autostart_in) do
		if mp.get_property("path"):lower():find(pp:lower()) or mp.get_property("working-directory"):lower():find(pp:lower()) then
			s1()
			break
		end
	end
end

function s1_1()
	if running == true then
		s_rm()
		mp.command('show-text "Quitting interSubs..."')
	else
		if mp.get_property("sub") == 'no' then
			mp.command('show-text "Select subtitles before starting interSubs."')
		else
			s1()
			mp.command('show-text "Starting interSubs..."')
		end
	end
end

function hide_show()
	if running == true then
		if hidden == true then
			os.execute('rm "' .. mpv_socket_2 .. '_hide" &')
			mp.osd_message("Showing interSubs.", .8)
			hidden = false
		else
			os.execute('touch "' .. mpv_socket_2 .. '_hide" &')
			mp.osd_message("Hiding interSubs.", .8)
			hidden = true
		end
	end
end

mp.add_forced_key_binding(keybinding, "start-stop-interSubs", s1_1)
mp.add_forced_key_binding(keybinding_hide, "hide-show-interSubs", hide_show)
mp.register_event("file-loaded", started)


Tk/interSubs.conf.py
# v. 1.23
# Interactive subtitles for `mpv` for language learners.

# BEWARE!
# treat this config like python
# next line shows how the main scripts gets these variables
# exec(open('interSubs.conf.py').read())

#########################################

# make sure selected translation function supports your language and so that codes of your languages are correct
# for instance Pons doesn't support Hebrew, Google and Reverso do, but their codes are different: 'iw' and 'he' respectively
# translate from language
lang_from = 'de'
# translate to language
lang_to = 'en'

############
# all *_B variables are boolean
############

pause_during_translation_B = True
# don't hide subtitle when its time is up and keep it on screen until the next line
extend_subs_duration2max_B = True
# show interSubs only in fullscreen
hide_when_not_fullscreen_B = True
# saving to ~/.config/mpv/scripts/urls/
save_translations_B = True
# every translation but first will be shuffled # scrolling through transitions would be disabled
randomize_translations_B = False

############################################

# translated language is written right-to-left, e.g Hebrew/Arabic
R2L_from_B = False
# translation is written right-to-left, e.g Hebrew/Arabic
R2L_to_B = False

font1 = ("Trebuchet MS", 33)		# subtitles (font, size)
font2 = ("Trebuchet MS", 27)		# [popup] original language & translation
font3 = ("Trebuchet MS", 23)		# [popup] morphology
font_color1 = '#BAC4D6'				# subtitles
font_color2 = '#DCDCCC'				# [popup] original language
font_color3 = '#8B8F88'				# [popup] translation
font_color4 = '#CA8200'				# [popup] morphology
font_color5 = '#1E90FF'				# [popup] nouns, masculine
font_color6 = '#BD3030'				# [popup] nouns, feminine
font_color7 = '#6DB56D'				# [popup] nouns, neuter

font_color8 = '#5EB0FF'				# [colorize_nouns] nouns, masculine
font_color9 = '#FFA6B5'				# [colorize_nouns] nouns, feminine
font_color10 = '#90EE90'			# [colorize_nouns] nouns, neuter

bg_color1 = '#000000'				# subtitles
bg_color2 = '#2C2C2C'				# translation popup

# show subtitles at the top of the screen
subs_top_placement_B = False
subs_screen_edge_padding = 5
popup_ext_n_int_padding = 5

##################################################

# number of translations in popup
number_of_translations = 3
# number of translations to save in files for each word; 0 - to save all
number_of_translations_to_save = 0

# interval in seconds between checking for the next subtitle
update_time = .01
# interval in seconds between checking if mpv is in focus using `xdotool` and/or in fullscreen
focus_checking_time = .1

# firefox "https://en.wiktionary.org/wiki/${word}"
show_in_browser = 'chromium "http://www.linguee.com/german-english/search?source=german&query=${word}"'

# dictionaries to use, one or more
# or other function's name you might write that will return ([[word, translation]..], [morphology = '', gender = ''])
# available:
#	pons
#	reverso
#	dict_cc
#	mtranslate_google (one word translation - for uncommon languages)
#	redensarten (redensarten-index.de - German idioms etc.)
#	tab_divided_dict - offline dictionary with word \t translation
translation_function_names = ['dict_cc', 'pons']

# path to the offline dictionary
tab_divided_dict_fname = '~/d/python_shit/mpv/scripts/z'
# strip <.*?>
tab_divided_dict_remove_tags_B = True

# for going through lines step by step
# skip pausing when subs are less then X words
auto_pause_min_words = 10
# 0 - don't pause
# 1 - pause after subs change
# 2 - pause before subs change
# wheel click on interSubs cycles through options
auto_pause = 0

# when subtitle consists of only one overly long line - splitting into two
split_long_lines_B = True
# split when there are more than N words in line
split_long_lines_words_min = 7
# colorize nouns by gender; German only with given dictionary
colorize_nouns_B = False

# it should have pattern: Word\t\tGender\t...
# exact genders are Masc/Fem/Neut
colorization_dict = 'interSubs.delexicon.txt'

# gtts|pons|forvo # gtts is google-text-to-speech
listen_via = 'gtts'

# filename where to save words if needed by bound to mouse-button function f_save_word_to_file; checks if the word is already there.
save_word_to_file_fname = '~/saved_words_by_interSubs'

# number of previous subtitles to show [0..inf]
show_N_of_previous_subtitles = 0
# show prev. subs above or below current; active if show_N_of_previous_subtitles > 0
show_previous_subtitles_above_current_B = True

###################################################
###################################################
# reassigning mouse buttons #
###################################################
###################################################
# functions' names are self-explanatory
# [mouse_event, modifier_key, 'self_explanatory_function_name']

# to one button multiple functions can be assigned
# by left-click this will open word in browser and save it to file.
# [1, 0, 'f_show_in_browser']
# [1, 0, 'f_save_word_to_file']

# mouse_event:
	# 1 == left-click
	# 2 == middle-click(wheel)
	# 3 == right-click
	# 4 == wheel down
	# 5 == wheel up
	# 6 == [non-standart] wheel left-click
	# 7 == [non-standart] wheel right-click
	# 8 == [non-standart] forward-button
	# 9 == [non-standart] back-button

# modifier_key:
	# 0 == None
	# 1 == Shift
	# 4 == Ctrl

# self_explanatory_function_name:
	# f_show_in_browser
	# f_auto_pause_options
	# f_listen
	# f_scroll_translations_down
	# f_scroll_translations_up
	# f_subs_screen_edge_padding_decrease
	# f_subs_screen_edge_padding_increase
	# f_font_size_decrease
	# f_font_size_increase
	# f_auto_pause_min_words_decrease
	# f_auto_pause_min_words_increase
	# f_deepl_translation
	# f_save_word_to_file

mouse_buttons = [
	[1, 0, 'f_show_in_browser'],
	[2, 0, 'f_auto_pause_options'],
	[3, 0, 'f_listen'],

	[4, 0, 'f_scroll_translations_down'],
	[5, 0, 'f_scroll_translations_up'],

	[4, 1, 'f_subs_screen_edge_padding_decrease'],
	[5, 1, 'f_subs_screen_edge_padding_increase'],

	[4, 4, 'f_font_size_decrease'],
	[5, 4, 'f_font_size_increase'],

	[6, 0, 'f_auto_pause_min_words_decrease'],
	[7, 0, 'f_auto_pause_min_words_increase'],
	[8, 0, 'f_deepl_translation'],

	[9, 0, ''],
]


################################
# below backwards compatibility stuff
################################
pause_during_translation = pause_during_translation_B
extend_subs_duration2max = extend_subs_duration2max_B
hide_when_not_fullscreen = hide_when_not_fullscreen_B
save_translations = save_translations_B
randomize_translations = randomize_translations_B
R2L_from = R2L_from_B
R2L_to = R2L_to_B
tab_divided_dict_remove_tags = tab_divided_dict_remove_tags_B
split_long_lines = split_long_lines_B
colorize_nouns = colorize_nouns_B
subs_bottom_padding = subs_screen_edge_padding


interSubs_config.py
#! /usr/bin/env python

# v. 2.9
# Interactive subtitles for `mpv` for language learners.

#########################################
# all *_B variables are boolean
######

# make sure selected translation function supports your language and so that codes of your languages are correct
# for instance Pons doesn't support Hebrew, Google and Reverso do, but their codes are different: 'iw' and 'he' respectively
# translate from language
lang_from = 'de'
# translate to language
lang_to = 'en'

# dictionaries to use, one or more
# or other function's name you might write that will return ([[word, translation]..], [morphology = '', gender = ''])
# available:
#	pons
#	reverso
#	dict_cc
#	leo
#	google
#	morfix (he/en en/he)
#	redensarten (redensarten-index.de - German idioms etc.)
#	tab_divided_dict - simple offline dictionary with word \t translation per line
translation_function_names = ['pons', 'reverso']
# for automatic switch to Hebrew. Skip if it isn't your language.
translation_function_names_2 = ['google', 'morfix']

# deepl/google
translation_function_name_full_sentence = 'google'

# number of translations in popup
number_of_translations = 4
# number of translations to save in files for each word; 0 - to save all
number_of_translations_to_save = 50

# gtts|pons|forvo # gtts is google-text-to-speech
listen_via = 'forvo'

# path to the offline dictionary
tab_divided_dict_fname = '~/d/python_shit/mpv/scripts/z.dict'
# strip <.*?>
tab_divided_dict_remove_tags_B = True

pause_during_translation_B = True
# don't hide subtitle when its time is up and keep it on screen until the next line
extend_subs_duration2max_B = True
# limit extension duration in seconds; N == 0: do not limit
extend_subs_duration_limit_sec = 33
# show interSubs only in fullscreen
hide_when_not_fullscreen_B = True

# interval between checking for the next subtitle; in seconds
update_time = .01
# interval in seconds between checking if mpv is in focus using `xdotool` and/or in fullscreen; in seconds
focus_checking_time = .1

# firefox "https://en.wiktionary.org/wiki/${word}"
show_in_browser = 'chromium "http://www.linguee.com/german-english/search?source=german&query=${word}"'

# filename where to save words if needed by bound to mouse-button function f_save_word_to_file; checks if the word is already there.
save_word_to_file_fname = '~/saved_words_by_interSubs'

# for going through lines step by step
# skip pausing when subs are less then X words
auto_pause_min_words = 10
# 0 - don't pause
# 1 - pause after subs change
# 2 - pause before subs change
# wheel click on interSubs cycles through options
auto_pause = 0

# translated language is written right-to-left, e.g Hebrew/Arabic
R2L_from_B = False
# translation is written right-to-left, e.g Hebrew/Arabic
R2L_to_B = False

#########################################################
#########################################################
### LOOKS ###
#########################################################
#########################################################

# show subtitles at the top of the screen
subs_top_placement_B = False
# distance to the edge; in px
subs_screen_edge_padding = 1
subs_padding_between_lines = 0

# when subtitle consists of only one overly long line - splitting into two
split_long_lines_B = True
# split when there are more than N words in line
split_long_lines_words_min = 8

# Qt's line wrapping doesn't work well
split_long_lines_in_popup_B = True
# split when there are more than N symbols in given line
split_long_lines_in_popup_symbols_min = 80

# ~ http://doc.qt.io/qt-5/stylesheet-reference.html
'''
/* Examples of css: */

background: transparent;				/* fully transparent */
background: black;						/* black background*/
background: #ffffff;					/* white background*/
background: rgba(0, 0, 0, 20%);			/* semi-opaque black 20% */
background: rgba(0, 0, 0, 50%);			/* semi-opaque black 50% */
background: rgba(0, 0, 0, 80%);			/* semi-opaque black 80% */
background: rgba(255, 255, 255, 40%);	/* semi-opaque white 40% */
background: rgba(44, 44, 44, 90%);		/* semi-opaque dark-grey 90% */

/* Font colors: */
color: white;
color: #BAC4D6;
color: rgb(217, 49, 49);				/* red */
color: rgba(217, 49, 49, 70%);			/* semi-opaque red 70% */

font-family: "Trebuchet MS";
font-weight: bold;
font-size: 33px;

font: bold italic large "Times New Roman" 34px;

		font-family: "FiraGO";
		font-family: "Trebuchet MS";
		font-family: "American Typewriter";
'''

# CSS styles for subtitles
style_subs = '''
	/* looks of subtitles */
	QFrame {
		background: transparent;
		color: white;
		color: #FFF0CD;

		font-family: "American Typewriter";
		/* font-weight: bold; */
		font-size: 52px;
	}
'''

# CSS styles for translations(popup)
style_popup = '''
	/* main */
	QFrame {
		background: rgb(44, 44, 44);

		font-family: "Trebuchet MS";
		font-weight: bold;
		font-size: 40px;
	}
	/* original language */
	QLabel#first_line {
		color: #DCDCCC;
	}

	/* original language - underlining exact word */
	QLabel#first_line_emphasize_word {
		color: #DCDCCC;
		text-decoration: underline;
	}

	/* translation */
	QLabel#second_line {
		color: #8B8F88;
	}

	/* colorizing morphology */
	[morphology=""]		{ color: #CA8200; }
	[morphology="m"]	{ color: #5EB0FF; }
	[morphology="f"]	{ color: #E34840; }
	[morphology="nt"]	{ color: #8BC34A; }

	/* delimiter between dictionaries */
	QFrame#delimiter {
		background: #8B8F88;
		font-size: 4px;	/* emulating thickness */
	}
'''

# for subtitles to be visible on background with similar color
# might look ugly with some fonts
outline_B = True
outline_color = '#000000'
# ~ N/2.5 == size in px; here it's 2px of outline + another ~ 3px of blur
outline_thickness = 5
outline_blur = 7
# change if outline is cropped from top/bottom of some letters depending on font
# in px; can take negative values
outline_top_padding = -2
outline_bottom_padding = 2

# highlighting the word under cursor
hover_color = '#F44336'
hover_hightlight = False	# may look ugly due to only int precision of QFontMetrics
hover_underline = True
hover_underline_thickness = 5

#########################################################
#########################################################
### reassigning mouse buttons ###
#########################################################
#########################################################

# functions' names are self-explanatory
# ['mouse_event', 'modifier_key', 'self_explanatory_function_name'],

# to one button multiple functions can be assigned
# by left-click this will open word in browser and save it to file.
# ['LeftButton', 'NoModifier', 'f_show_in_browser'],
# ['LeftButton', 'NoModifier', 'f_save_word_to_file'],

# https://doc.qt.io/qt-5/qt.html#MouseButton-enum
# mouse_event:
	# LeftButton
	# RightButton
	# MiddleButton (wheel-click)
	# BackButton (Typically present on the 'thumb' side of a mouse with extra buttons. This is NOT the tilt wheel.)
	# ForwardButton

	# wheel scroll up/down left/right names' are arbitrary and not from Qt
	# ScrollUp
	# ScrollDown
	# ScrollLeft (This is the tilt wheel.)
	# ScrollRight

# Note: On macOS, the ControlModifier value corresponds to the Command keys on the keyboard.
# https://doc.qt.io/qt-5/qt.html#KeyboardModifier-enum
# modifier_key:
	# NoModifier
	# ControlModifier
	# ShiftModifier
	# AltModifier

# self_explanatory_function_name:
	# f_show_in_browser
	# f_auto_pause_options
	# f_listen
	# f_scroll_translations_down
	# f_scroll_translations_up
	# f_subs_screen_edge_padding_decrease
	# f_subs_screen_edge_padding_increase
	# f_font_size_decrease
	# f_font_size_increase
	# f_auto_pause_min_words_decrease
	# f_auto_pause_min_words_increase
	# f_translation_full_sentence
	# f_save_word_to_file
	# f_deepl_translation < obsolete, changed into f_translation_full_sentence

mouse_buttons = [
	['LeftButton',		'NoModifier',		'f_show_in_browser'],
	['RightButton',		'NoModifier',		'f_translation_full_sentence'],
	['MiddleButton',	'NoModifier',		'f_auto_pause_options'],

	['BackButton',		'NoModifier',		'f_listen'],

	['ScrollDown',		'NoModifier',		'f_scroll_translations_down'],
	['ScrollUp',		'NoModifier',		'f_scroll_translations_up'],

	['ScrollDown',		'ControlModifier',		'f_font_size_decrease'],
	['ScrollUp',		'ControlModifier',		'f_font_size_increase'],

	# ['ScrollLeft',		'NoModifier',		'f_auto_pause_min_words_decrease'],
	# ['ScrollRight',		'NoModifier',		'f_auto_pause_min_words_increase'],
	['ScrollRight',		'NoModifier',		'f_listen'],

	['ScrollDown',		'ShiftModifier',		'f_subs_screen_edge_padding_decrease'],
	['ScrollUp',		'ShiftModifier',		'f_subs_screen_edge_padding_increase'],
]

# http://culmus.sourceforge.net/summary.html
he_fonts = ['Miriam', 'Miriam Mono', 'Drugulin', 'David', 'Frank Ruehl', 'Shofar', 'Varela Round', 'FiraGO']

# obsolete vars
hover_underline_width = hover_underline_thickness
